File: e2e/kcp/zig/src/main.zig
//! KCP stream interoperability test — Zig side.
//!
//! Usage: kcp_test --name zig --config ../config.json

const std = @import("std");
const noise = @import("noise");

const Key = noise.Key;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Parse args.
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    var name: ?[]const u8 = null;
    var config_path: ?[]const u8 = null;
    var i: usize = 1;
    while (i < args.len) : (i += 1) {
        if (std.mem.eql(u8, args[i], "--name") and i + 1 < args.len) {
            i += 1;
            name = args[i];
        } else if (std.mem.eql(u8, args[i], "--config") and i + 1 < args.len) {
            i += 1;
            config_path = args[i];
        }
    }
    const my_name = name orelse return error.MissingName;
    const cfg_path = config_path orelse return error.MissingConfig;

    // Parse config JSON.
    const config_data = try std.fs.cwd().readFileAlloc(allocator, cfg_path, 1024 * 1024);
    defer allocator.free(config_data);

    const parsed = try std.json.parseFromSlice(Config, allocator, config_data, .{ .ignore_unknown_fields = true });
    defer parsed.deinit();
    const config = parsed.value;

    // Find our host.
    var my_host: ?HostInfo = null;
    var peer_host: ?HostInfo = null;
    for (config.hosts) |h| {
        if (std.mem.eql(u8, h.name, my_name)) {
            my_host = h;
        } else {
            peer_host = h;
        }
    }
    const me = my_host orelse return error.HostNotFound;
    const peer = peer_host orelse return error.PeerNotFound;

    // Parse keys.
    var priv_key: [32]u8 = undefined;
    _ = try std.fmt.hexToBytes(&priv_key, me.private_key);
    const kp = noise.KeyPair.fromPrivate(Key.fromBytes(priv_key));

    var peer_priv: [32]u8 = undefined;
    _ = try std.fmt.hexToBytes(&peer_priv, peer.private_key);
    const peer_kp = noise.KeyPair.fromPrivate(Key.fromBytes(peer_priv));

    std.log.info("[{s}] Public key: {x}...", .{ my_name, kp.public.asBytes()[0..8].* });
    std.log.info("[{s}] Role: {s}", .{ my_name, me.role });

    // Create UDP.
    const bind_addr_str = try std.fmt.allocPrint(allocator, "127.0.0.1:{d}", .{me.port});
    defer allocator.free(bind_addr_str);

    const udp = try noise.UDP.init(allocator, &kp, .{
        .bind_addr = bind_addr_str,
        .allow_unknown = true,
        .decrypt_workers = 1,
    });
    defer udp.deinit();

    std.log.info("[{s}] Listening on port {d}", .{ my_name, me.port });

    // Add peer.
    const Endpoint = noise.net.endpoint_mod.Endpoint;
    udp.setPeerEndpoint(peer_kp.public, Endpoint.init(.{ 127, 0, 0, 1 }, peer.port));

    // Determine mode.
    const mode = config.@"test".mode;
    const is_opener = std.mem.eql(u8, me.role, "opener");

    if (is_opener) {
        std.log.info("[{s}] Waiting for peer...", .{my_name});
        std.Thread.sleep(2 * std.time.ns_per_s);

        std.log.info("[{s}] Connecting...", .{my_name});
        try udp.connectTimeout(&peer_kp.public, 10_000);
        std.Thread.sleep(200 * std.time.ns_per_ms);

        // Open stream.
        var stream = try udp.openStream(&peer_kp.public, 1); // Service.proxy = 1
        defer stream.close();

        if (std.mem.eql(u8, mode, "echo") or mode.len == 0) {
            const msg = config.@"test".echo_message;
            _ = try stream.write(msg);
            std.log.info("[opener] Sent: {s}", .{msg});
            var buf: [4096]u8 = undefined;
            const n = try stream.read(&buf);
            std.log.info("[opener] Response: {s}", .{buf[0..n]});
        } else if (std.mem.eql(u8, mode, "streaming")) {
            const total = config.@"test".throughput_mb * 1024 * 1024;
            const chunk_size = if (config.@"test".chunk_kb > 0) config.@"test".chunk_kb * 1024 else 8192;
            const chunk = try allocator.alloc(u8, chunk_size);
            defer allocator.free(chunk);
            @memset(chunk, 0x58);
            var sent: usize = 0;
            while (sent < total) {
                const n = try stream.write(chunk);
                sent += n;
            }
            std.log.info("[opener] Sent {d} bytes", .{sent});
        } else if (std.mem.eql(u8, mode, "multi_stream")) {
            const num_streams = if (config.@"test".num_streams > 0) config.@"test".num_streams else 10;
            const data = try allocator.alloc(u8, 100 * 1024);
            defer allocator.free(data);
            for (data, 0..) |*b, idx| b.* = @intCast(idx % 256);

            // First stream already opened above.
            _ = try stream.write(data);
            std.log.info("[opener] stream 0: sent {d} bytes", .{data.len});

            var i_stream: usize = 1;
            while (i_stream < num_streams) : (i_stream += 1) {
                var s = try udp.openStream(&peer_kp.public, 1);
                defer s.close();
                _ = try s.write(data);
                std.log.info("[opener] stream {d}: sent {d} bytes", .{ i_stream, data.len });
            }
            std.log.info("[opener] all {d} streams done", .{num_streams});
        } else if (std.mem.eql(u8, mode, "delayed_write")) {
            const delay_ms = if (config.@"test".delay_ms > 0) config.@"test".delay_ms else 2000;
            std.log.info("[opener] delaying {d}ms before writing...", .{delay_ms});
            std.Thread.sleep(delay_ms * std.time.ns_per_ms);
            _ = try stream.write("delayed hello");

            var buf: [4096]u8 = undefined;
            const n = try stream.read(&buf);
            std.log.info("[opener] delayed response: {s}", .{buf[0..n]});
        } else {
            std.log.err("unknown mode: {s}", .{mode});
            return error.InvalidMode;
        }

        std.Thread.sleep(1 * std.time.ns_per_s);
    } else {
        std.log.info("[{s}] Waiting for connection...", .{my_name});

        // Wait for opener to connect.
        std.Thread.sleep(5 * std.time.ns_per_s);

        const result = try udp.acceptStream(&peer_kp.public);
        var stream = result.stream;
        defer stream.close();

        std.log.info("[accepter] Accepted stream on service={d}", .{result.service});

        if (std.mem.eql(u8, mode, "echo") or mode.len == 0 or std.mem.eql(u8, mode, "delayed_write")) {
            var buf: [4096]u8 = undefined;
            const n = try stream.read(&buf);
            std.log.info("[accepter] Received: {s}", .{buf[0..n]});
            const response = try std.fmt.allocPrint(allocator, "Echo: {s}", .{buf[0..n]});
            defer allocator.free(response);
            _ = try stream.write(response);
            std.log.info("[accepter] Sent: {s}", .{response});
        } else if (std.mem.eql(u8, mode, "streaming")) {
            const total = config.@"test".throughput_mb * 1024 * 1024;
            var buf: [65536]u8 = undefined;
            var recv: usize = 0;
            var last_log: usize = 0;
            while (recv < total) {
                const n = try stream.read(&buf);
                if (n == 0) break;
                recv += n;
                if (recv - last_log >= 256 * 1024) {
                    std.log.info("[accepter] progress {d} / {d}", .{ recv, total });
                    last_log = recv;
                }
            }
            std.log.info("[accepter] Received {d} / {d} bytes", .{ recv, total });
            if (recv < total) return error.IncompleteTransfer;
        } else if (std.mem.eql(u8, mode, "multi_stream")) {
            const num_streams = if (config.@"test".num_streams > 0) config.@"test".num_streams else 10;
            const expected_per_stream: usize = 100 * 1024;

            // First stream already accepted above.
            {
                var buf: [65536]u8 = undefined;
                var recv: usize = 0;
                while (recv < expected_per_stream) {
                    const n = try stream.read(&buf);
                    if (n == 0) break;
                    recv += n;
                }
                std.log.info("[accepter] stream 0: received {d} bytes", .{recv});
            }

            var stream_idx: usize = 1;
            while (stream_idx < num_streams) : (stream_idx += 1) {
                const r = try udp.acceptStream(&peer_kp.public);
                var s = r.stream;
                defer s.close();

                var buf: [65536]u8 = undefined;
                var recv: usize = 0;
                while (recv < expected_per_stream) {
                    const n = try s.read(&buf);
                    if (n == 0) break;
                    recv += n;
                }
                std.log.info("[accepter] stream {d}: received {d} bytes", .{ stream_idx, recv });
            }
            std.log.info("[accepter] all {d} streams done", .{num_streams});
        } else {
            std.log.err("unknown mode: {s}", .{mode});
            return error.InvalidMode;
        }

        std.Thread.sleep(1 * std.time.ns_per_s);
    }

    std.log.info("[{s}] Test completed successfully!", .{my_name});
}

const Config = struct {
    hosts: []HostInfo,
    @"test": TestConfig,
};

const HostInfo = struct {
    name: []const u8,
    private_key: []const u8,
    port: u16,
    role: []const u8,
};

const TestConfig = struct {
    mode: []const u8 = "",
    echo_message: []const u8 = "",
    throughput_mb: usize = 0,
    chunk_kb: usize = 0,
    num_streams: usize = 0,
    delay_ms: u64 = 0,
};

File: e2e/kcp/rust/src/main.rs
//! KCP stream interoperability test — Rust side.
//!
//! Usage: kcp_test --name rust --config ../config.json

use std::env;
use std::fs;
use std::io::{Read, Write};
use std::net::SocketAddr;
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

use zgrnet::noise::message::service;
use zgrnet::{Key, KeyPair, SyncStream};
use zgrnet::{UDP, UdpOptions};

#[derive(serde::Deserialize)]
struct Config {
    hosts: Vec<HostInfo>,
    test: TestConfig,
}

#[derive(serde::Deserialize)]
struct HostInfo {
    name: String,
    private_key: String,
    port: u16,
    role: String,
}

#[derive(serde::Deserialize)]
struct TestConfig {
    #[serde(default)]
    mode: String,
    #[serde(default)]
    echo_message: String,
    #[serde(default)]
    throughput_mb: usize,
    #[serde(default)]
    chunk_kb: usize,
    #[serde(default)]
    num_streams: usize,
    #[serde(default)]
    delay_ms: u64,
}

fn main() {
    let rt = tokio::runtime::Builder::new_multi_thread()
        .worker_threads(4)
        .enable_all()
        .build()
        .expect("tokio runtime");
    // Run everything inside the runtime so all threads have access.
    rt.block_on(async { tokio::task::spawn_blocking(run).await.unwrap() })
}

fn run() {

    let args: Vec<String> = env::args().collect();
    let name = args.iter().position(|x| x == "--name")
        .and_then(|i| args.get(i + 1)).map(|s| s.as_str())
        .expect("Usage: --name <name> --config <path>");
    let config_path = args.iter().position(|x| x == "--config")
        .and_then(|i| args.get(i + 1)).map(|s| s.to_string())
        .expect("Usage: --name <name> --config <path>");

    let config: Config = serde_json::from_str(
        &fs::read_to_string(&config_path).unwrap()
    ).unwrap();

    let my_host = config.hosts.iter().find(|h| h.name == name)
        .unwrap_or_else(|| panic!("Host {} not found", name));

    let priv_bytes = hex::decode(&my_host.private_key).unwrap();
    let mut priv_key = [0u8; 32];
    priv_key.copy_from_slice(&priv_bytes);
    let key_pair = KeyPair::from_private(Key::from(priv_key));

    eprintln!("[{}] Public key: {}...", name, hex::encode(&key_pair.public.as_bytes()[..8]));
    eprintln!("[{}] Role: {}", name, my_host.role);

    let bind_addr = format!("0.0.0.0:{}", my_host.port);
    let udp = Arc::new(UDP::new(
        key_pair.clone(),
        UdpOptions::new().bind_addr(&bind_addr).allow_unknown(true),
    ).expect("Failed to create UDP"));

    eprintln!("[{}] Listening on {}", name, udp.host_info().addr);

    let peer_host = config.hosts.iter().find(|h| h.name != name)
        .expect("No peer found");

    let peer_priv = hex::decode(&peer_host.private_key).unwrap();
    let mut pp = [0u8; 32];
    pp.copy_from_slice(&peer_priv);
    let peer_kp = KeyPair::from_private(Key::from(pp));

    let endpoint: SocketAddr = format!("127.0.0.1:{}", peer_host.port).parse().unwrap();
    udp.set_peer_endpoint(peer_kp.public, endpoint);

    let udp_recv = Arc::clone(&udp);
    thread::spawn(move || {
        let mut buf = vec![0u8; 65535];
        loop {
            if udp_recv.is_closed() { break; }
            let _ = udp_recv.read_from(&mut buf);
        }
    });

    if my_host.role == "opener" {
        eprintln!("[{}] Waiting for peer...", name);
        thread::sleep(Duration::from_secs(2));

        eprintln!("[{}] Connecting...", name);
        udp.connect(&peer_kp.public).expect("connect failed");

        let mut established = false;
        for i in 0..100 {
            if let Some(info) = udp.peer_info(&peer_kp.public) {
                eprintln!("[{}] peer state ({}): {:?}", name, i, info.state);
                if info.state == zgrnet::PeerState::Established {
                    established = true;
                    break;
                }
            }
            thread::sleep(Duration::from_millis(50));
        }
        if !established {
            panic!("session never established after 5 seconds");
        }
        thread::sleep(Duration::from_millis(200));

        run_opener(&udp, &peer_kp.public, &config.test);
    } else {
        eprintln!("[{}] Waiting for connection...", name);
        run_accepter(&udp, &peer_kp.public, &config.test);
    }

    eprintln!("[{}] Test completed successfully!", name);
    udp.close().unwrap();
}

fn run_opener(udp: &Arc<UDP>, peer_pk: &Key, cfg: &TestConfig) {
    let mode = if cfg.mode.is_empty() { "echo" } else { &cfg.mode };

    let mut stream = udp.open_stream(peer_pk, service::PROXY)
        .expect("open_stream failed");
    eprintln!("[opener] Opened stream (mode={})", mode);

    match mode {
        "echo" => {
            stream.write_all(cfg.echo_message.as_bytes()).unwrap();
            eprintln!("[opener] Sent: {:?}", cfg.echo_message);
            let mut buf = vec![0u8; 4096];
            let n = read_with_timeout(&mut stream, &mut buf, Duration::from_secs(10))
                .expect("echo read failed");
            eprintln!("[opener] Response: {:?}", String::from_utf8_lossy(&buf[..n]));
        }
        "streaming" => {
            let total = cfg.throughput_mb * 1024 * 1024;
            let chunk_size = if cfg.chunk_kb > 0 { cfg.chunk_kb * 1024 } else { 8192 };
            let chunk: Vec<u8> = (0..chunk_size).map(|i| (i % 256) as u8).collect();
            let mut sent = 0;
            while sent < total {
                let n = stream.write(&chunk).unwrap();
                sent += n;
            }
            eprintln!("[opener] Sent {} bytes", sent);
        }
        "multi_stream" => {
            let num = if cfg.num_streams > 0 { cfg.num_streams } else { 10 };
            let data: Vec<u8> = (0..100 * 1024).map(|i| (i % 256) as u8).collect();
            stream.write_all(&data).unwrap();

            let mut handles = Vec::new();
            for i in 1..num {
                let mut s = udp.open_stream(peer_pk, service::PROXY)
                    .unwrap_or_else(|e| panic!("open stream {}: {}", i, e));
                let data = data.clone();
                handles.push(thread::spawn(move || {
                    s.write_all(&data).unwrap();
                    eprintln!("[opener] stream {}: sent {} bytes", i, data.len());
                    s.close();
                }));
            }
            for h in handles { h.join().unwrap(); }
            eprintln!("[opener] all {} streams done", num);
        }
        "delayed_write" => {
            let delay = if cfg.delay_ms > 0 { cfg.delay_ms } else { 2000 };
            eprintln!("[opener] delaying {}ms...", delay);
            thread::sleep(Duration::from_millis(delay));
            stream.write_all(b"delayed hello").unwrap();
            let mut buf = vec![0u8; 4096];
            let n = read_with_timeout(&mut stream, &mut buf, Duration::from_secs(10))
                .expect("delayed read failed");
            eprintln!("[opener] delayed response: {:?}", String::from_utf8_lossy(&buf[..n]));
        }
        _ => panic!("unknown mode: {}", mode),
    }

    thread::sleep(Duration::from_secs(1));
    stream.close();
}

fn run_accepter(udp: &Arc<UDP>, peer_pk: &Key, cfg: &TestConfig) {
    let deadline = Instant::now() + Duration::from_secs(30);
    loop {
        if Instant::now() > deadline { panic!("timeout waiting for session"); }
        if let Some(info) = udp.peer_info(peer_pk) {
            if info.state == zgrnet::PeerState::Established { break; }
        }
        thread::sleep(Duration::from_millis(100));
    }
    thread::sleep(Duration::from_millis(100));

    let mode = if cfg.mode.is_empty() { "echo" } else { &cfg.mode };

    match mode {
        "echo" | "delayed_write" => {
            let (mut stream, _svc) = udp.accept_stream(peer_pk).expect("accept failed");
            let mut buf = vec![0u8; 4096];
            let n = read_with_timeout(&mut stream, &mut buf, Duration::from_secs(30))
                .expect("read failed");
            eprintln!("[accepter] Received: {:?}", String::from_utf8_lossy(&buf[..n]));
            let response = format!("Echo: {}", String::from_utf8_lossy(&buf[..n]));
            stream.write_all(response.as_bytes()).unwrap();
            stream.close();
        }
        "streaming" => {
            let (mut stream, _svc) = udp.accept_stream(peer_pk).expect("accept failed");
            let total = cfg.throughput_mb * 1024 * 1024;
            let mut buf = vec![0u8; 65536];
            let mut recv = 0;
            while recv < total {
                match stream.read(&mut buf) {
                    Ok(0) => break,
                    Ok(n) => recv += n,
                    Err(_) => break,
                }
            }
            eprintln!("[accepter] Received {} / {} bytes", recv, total);
            if recv < total { panic!("incomplete: {} < {}", recv, total); }
            stream.close();
        }
        "multi_stream" => {
            let num = if cfg.num_streams > 0 { cfg.num_streams } else { 10 };
            let mut handles = Vec::new();
            for i in 0..num {
                let (mut stream, _svc) = udp.accept_stream(peer_pk)
                    .unwrap_or_else(|e| panic!("accept {}: {}", i, e));
                handles.push(thread::spawn(move || {
                    let mut buf = vec![0u8; 65536];
                    let mut total = 0;
                    loop {
                        match stream.read(&mut buf) {
                            Ok(0) => break,
                            Ok(n) => total += n,
                            Err(_) => break,
                        }
                    }
                    eprintln!("[accepter] stream {}: received {} bytes", i, total);
                    stream.close();
                }));
            }
            for h in handles { h.join().unwrap(); }
            eprintln!("[accepter] all {} streams done", num);
        }
        _ => panic!("unknown mode: {}", mode),
    }

    thread::sleep(Duration::from_secs(1));
}

fn read_with_timeout(stream: &mut SyncStream, buf: &mut [u8], timeout: Duration) -> Result<usize, String> {
    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        match stream.read(buf) {
            Ok(0) => thread::sleep(Duration::from_millis(1)),
            Ok(n) => return Ok(n),
            Err(e) => return Err(format!("{}", e)),
        }
    }
    Err("read timeout".into())
}
